### ISSUE TITLE
echo residual during single talk

### ISSUE THREAD

**Issue by heartQi** [richiejp/deepvqe-ggml#2]
Hi,
Base your code, I trained a model but some echo residuals could not be completely eliminated during the single talk even using the complete linear echo, can you give some suggestion to fix this issue, thanks

**Comment by richiejp** [comment]
How much reduction (ERLE) or echo residual is there? What's the align accuracy?

**Comment by heartQi** [comment]
Epoch 98: train_loss=0.0407, val_loss=0.0406, delay_acc=99.8%, erle=+19.4dB, lr=1.00e-03
  single_talk_farend: erle=+48.3dB (n=1000), echo_mos=3.97, deg_mos=5.00, dnsmos=2.02
  double_talk: erle=+12.1dB (n=4000), pesq=2.07, echo_mos=4.05, deg_mos=3.88, dnsmos=2.80

This is one of the training logs.

**Comment by richiejp** [comment]
OK, I have been doing some more testing and it is completely failing to estimate the delay and suppress echo on some devices. I think the problem is that the simple distortions applied to the reference signal when synthesising the training data aren't realistic enough. So when it get's real audio it may work in some cases, but in others it is failing completely.

Using the extra training set from #1 didn't help in LocalVQE, I haven't tried it with DeepVQE. The first thing I'll try is improving the data synthesis pipeline to try to make the model generalise more.

**Comment by richiejp** [comment]
The training data doesn't cover a whole load of scenarios e.g. when the reference input is much quieter than the microphone. Also I didn't handle #1 correctly, to really use that data we need self-supervised learning. I'm implementing and testing a new data synthesis pipeline in LocalVQE because the model is much smaller making testing it faster (althoug I may have to increase the size after this xD), but then I'll back port to deepvqe.

**Comment by richiejp** [comment]
I have backported some changes I made to LocalVQE: https://github.com/richiejp/deepvqe-ggml/commit/dfd47a65d75586c3e2c9c08a9dec22e6bdd692de

My evaluations for LocalVQE showed that these make the AEC far more robust and it works perfectly on the real device I have tested on so far where previously it was broken. I haven't done a full training run on deepvqe-ggml yet though due to lack of GPUs.

**Comment by heartQi** [comment]
Thank you for your update. In addition to the robustness issue you described, the problem I've been trying to solve is that residual echoes are always present when dealing with linear echoes, and the peak is around -60dB (the delay estimate is correct).

<img width="2388" height="1872" alt="Image" src="https://github.com/user-attachments/assets/47c69b8c-3908-4de7-b50a-91c99cec7c98" />

In addition, I made an interesting test using the first version of the data synthesis method:
1. Using the same loss configuration:    mag_l1_weight: 1.0， time_l1_weight: 1.0
2. Setting single_talk_prob to 1.0 which means all training data are farend single talk

Test results show that deepvqe-ggml can not completely eliminate echoes and leak residual echoes. However, LocalVQE can completely eliminate the echoes. So I'm wondering if this is due to limitations in the implementation of the deepvqe-ggml model.

**Comment by richiejp** [comment]
Possibly there is an issue with the SOAP optimizer and switching to Schedule free AdamW or AdamW with sin annealing would allow it to converge on complete elimination. I dropped SOAP in LocalVQE because it kept driving some values to infinity and crashing training after switching the bottleneck from GRU to S4D and while SOAP is faster and sounds great I'm not sure it's worth fixing. Claude thinks that DeepVQE just needs more epochs especially because it is larger than LocalVQE. Reducing the data diversity may make performance worse as well if it locks onto some local minima. I did some over-training tests in the past and it didn't perfect those scenarios, but that needs to be taken with a pinch of salt because I was trying to shake out bugs at the time.

I'm a little surprised that LocalVQE completely eliminates residual echo, I thought there were always some artefacts left over at least which I assumed could just be removed with a noise gate if that is needed, but it's great if it does remove it completely. I was thinking that if it's too quiet to hear or for ASR to pick up, then that is good enough, or are there cases where it is not?

**Comment by richiejp** [comment]
I did a 23 epoch training run and have uploaded the weights: https://huggingface.co/richiejp/deepvqe-aec-gguf/tree/main. Epoch 21 reached 49.2 dB erle on single-talk which is less than LocalVQE, but I need the GPU for other things. However it was still progressing so it would get to the 51-54 dB that LocalVQE achieved.

**Comment by heartQi** [comment]
Test the epcoh_23 and still have the echo residual issue, which schedule and loss do you have used when you trained the 23 epoch.
I can try the Schedule free AdamW or AdamW with sin annealing to retrain deepvqe-ggml.

**Comment by thinkingwanglin** [comment]
I'm encountering the same issue during far-end single-talk scenarios. After testing, the enhanced audio achieves an ERLE of 47.98 dB, but it sounds and looks like the echo signal has merely been attenuated by a fixed ratio—the overall waveform still closely resembles the original echo.
Could this behavior be related to the loss function or optimizer choice? I’d really appreciate your insights on this. Thanks!

**Comment by thinkingwanglin** [comment]
Additionally, I noticed that during inference, the temperature parameter of the alignment module must be explicitly reset; otherwise, echo cancellation fails entirely. I suspect the same issue likely exists in LocalVQE when running the C++ inference module directly, though it may have gone unnoticed if the temperature was implicitly handled or hardcoded during export. This seems like an important detail for deployment consistency.

**Comment by richiejp** [comment]
> Test the epcoh_23 and still have the echo residual issue, which schedule and loss do you have used when you trained the 23 epoch. I can try the Schedule free AdamW or AdamW with sin annealing to retrain deepvqe-ggml.

https://github.com/richiejp/deepvqe-ggml/blob/6af6d18349d50e190e6493b50ffb625533d54187/train/configs/default.yaml

It's from this version.

**Comment by richiejp** [comment]
> Additionally, I noticed that during inference, the temperature parameter of the alignment module must be explicitly reset; otherwise, echo cancellation fails entirely. I suspect the same issue likely exists in LocalVQE when running the C++ inference module directly, though it may have gone unnoticed if the temperature was implicitly handled or hardcoded during export. This seems like an important detail for deployment consistency.

Correct on both accounts it seems! In LocalVQE the effect seems to be modest, except on some samples where it completely prevents AEC. In DeepVQE it seems to be quite bad also. I uploaded a fix so that the temp the model was trained with is included in the metadata and used at inference. So the temp was defaulting to 1.0, but should be 0.1 with the epoch I published. I don't know if this explains the lack of AEC you are seeing.

**Comment by richiejp** [comment]
> I'm encountering the same issue during far-end single-talk scenarios. After testing, the enhanced audio achieves an ERLE of 47.98 dB, but it sounds and looks like the echo signal has merely been attenuated by a fixed ratio—the overall waveform still closely resembles the original echo. Could this behavior be related to the loss function or optimizer choice? I’d really appreciate your insights on this. Thanks!

I can't rule out anything at this point including another bug. One thing I have been bitten by both in tensor board and gradio is that something was amplifying the audio to fill the dynamic range. So if you have an audio file with an almost silent residual it would get amplified into the audible range.

**Comment by thinkingwanglin** [comment]
> > I'm encountering the same issue during far-end single-talk scenarios. After testing, the enhanced audio achieves an ERLE of 47.98 dB, but it sounds and looks like the echo signal has merely been attenuated by a fixed ratio—the overall waveform still closely resembles the original echo. Could this behavior be related to the loss function or optimizer choice? I’d really appreciate your insights on this. Thanks!
> 
> I can't rule out anything at this point including another bug. One thing I have been bitten by both in tensor board and gradio is that something was amplifying the audio to fill the dynamic range. So if you have an audio file with an almost silent residual it would get amplified into the audible range.

Thanks for your reply!

I’ve tried increasing the proportion of far-end single-talk scenarios in the training data to 0.8. This has somewhat reduced the output volume of the enhanced audio, but the waveform still closely follows the original echo pattern. I suspect there might still be an issue with how the residual is being handled—possibly related to dynamic range normalization or scaling during visualization (as you mentioned with TensorBoard/Gradio).

I’ll keep investigating and adjusting the setup to pinpoint the root cause. Really appreciate your insight!
